### DISCUSSION TITLE
reset or copy?

### DISCUSSION THREAD

**Discussion by Tang115** [CURENT/andes#415] [General]
Hi Prof. Cui,

I am interested about is there any way to copy or reset an Andes model (TDS initialized). Either of the following scenarios is what I want to achieve:

1) suppose I run the TDS of a model until 10 seconds. I would like to reset this model while keeping all initialization conditions and re-run the model from second 0.
2) suppose I run the power flow of a model. I would like to copy the model and run the TDS on the copy one without affecting the original model. I already try copy.deepcopy(), but it returns: cannot pickle 'PyCapsule' object.

Looking forward to your comments.

Best

**Comment by cuihantao** [comment]
Hello,

It is not yet possible to reset a syste, but as you mentioned, you can copy a system that has just been initialized. See the snapshot function here: https://docs.andes.app/en/latest/_generated/andes.utils.snapshot.html

**Reply by Tang115** [reply]
It works for me. Thank you!

**Comment by MaKla89** [comment]
I've come to this point, too, now and while I was able to create a "deep-copy" via saving and loading the snapshot, I found a problem: This SEEMS to only work once. I use this:

`andes.utils.snapshot.save_ss("snapshot.pkl", andes_system)`
`panda_system = andes.utils.snapshot.load_ss("snapshot.pkl")`
`os.remove("snapshot.pkl")`

The first time this works without any problem. If I use that same code again, I get this error:

`RecursionError: maximum recursion depth exceeded while calling a Python object` for the first line (the saving part) of the code.

Is this something that obviously happens (at least if you know your way around pickle , etc.) or otherwise intendend behaviour? :) 


(Some context: In my case, I need to create a (deep) copy of the andes system in order to convert it into a pandapower system and run an OPF (if I use the original Andes system for that, the conversion-function changes the original system).  I need to run the OPF in pandapower twice at different times of the TDS. Hence, I want to create a temporary copy of the original Andes system twice, too.  Theoretically, the temporary copy is not needed any longer after the OPF is done)

**Reply by cuihantao** [reply]
It would be super helpful if you could provide a minimal example to reproduce. The error is not intended; it could be bugs in ANDES or dill, but I couldn't reproduce it with the code below. I performed save and load on a previously loaded snaoshot, and it works ok. 

```python

import andes
from matplotlib import pyplot

%matplotlib inline

andes.config_logger()

ss = andes.system.example()

ss.PFlow.run()
ss.TDS.init();

# --- First save and load ---

from andes.utils.snapshot import save_ss, load_ss

save_ss("snapshot.pkl", ss)

ss_new = load_ss("snapshot.pkl")

ss_new.TDS.run()

ss_new.TDS.plt.plot(ss.GENROU.omega)

# --- Second save and load ---
save_ss("snapshot.pkl", ss_new)

ss_new2 = load_ss("snapshot.pkl")

ss_new2.TDS.config.tf = 30
ss_new2.TDS.run()

ss_new2.TDS.plt.plot(ss.GENROU.omega)
```

**Reply by cuihantao** [reply]
Both of the `plt.plot` functions use `ss.GENROU.omega`, but it will give the same results if `ss` is replaced by `ss_new` or `ss_new2`.

**Reply by MaKla89** [reply]
Sorry, it took me a bit to create an isolated, concise example. The attached script does not run as `andes.utils.snapshot.save_ss("snapshot.pkl", gv_ps)` is executed a second time in line 150. This line throws  `RecursionError: maximum recursion depth exceeded while calling a Python object`. Removing the old pickle-file in line 149 was just an attempt, but does not change a thing.  If you comment out lines 149-151, the code runs nicely, using the snapshot feature ONCE (in lines 74-77). I do have the feeling that it is a general coding-basic rule that I violated in there (and I'm sorry if that is the case for wasting your time), but I do not see the problem.

[error_example.txt](https://github.com/cuihantao/andes/files/10911337/error_example.txt)

PS: Regarding your example the difference is that I'm not trying to CHAIN the calls but rather create a temporary copy the original system (in your case `ss`) at two different times. Simple example: I want to run the PP-OPF once during the system-initialisation and one more time after a disturbance happens. I cannot use the `to_pandapower`-function directly on the original system as the function changes the the used system. Therefore I need a (deep) copy, I think. As ANDES-systems cannot directly be pickled, I used the snapshot feature to fake a deep-copy. And it seems to be working, but just once :D

**Reply by cuihantao** [reply]
Thank you! Please allow me a few days to work on it.

**Comment by vaahedi** [comment]
Hello, has there been any update on the Recursive issue, we are getting that error when using snapshot loading as well

**Comment by cuihantao** [comment]
`deepcopy` will always have the issue because of the KLU `PyCapsule` object. Try calling `system.remove_pycapsule()` before copying.
